### ISSUE TITLE
ErrorProneInjector incorrectly picks up the no-args constructor

### ISSUE THREAD

**Issue by ksobolew** [google/error-prone#3931]
The logic of determining the constructor of a checker class changed in commit 1c3c09f5c286989f76eb545613c9ce91dac4acad. Previously it explicitly checked for a constructor with a single parameter of type `ErrorProneFlags`; now it looks for a constructor with all parameters having this type. The issue is that the condition used - `allMatch` - also returns `true` if there are no parameters at all, so the no-args constructor also matches the test.

The impact is that when I have a test for a checker which needs to take additional flags, the `CompilationTestHelper` uses the wrong constructor to create the checker instance, making it impossible to pass flags in a test.

**Comment by Stephan202** [comment]
@ksobolew this sounds similar to something that initially puzzled me when upgrading to Error Prone 2.19+. Then I realized that just annotating the `ErrorProneFlags`-accepting constructors with `@Inject` suffices. See the changes in PicnicSupermarket/error-prone-support#621 for an example. Hope this helps!

**Comment by Stephan202** [comment]
Check; I suppose that's something for @graememorgan to have an opinion on :)

**Comment by graememorgan** [comment]
Hah, I think I thought of that edge case when I wrote it, but decided it wouldn't matter for some reason. :)

I would expect the inject constructor to be annotated `@Inject` and therefore win.

> The impact is that when I have a test for a checker which needs to take additional flags, the CompilationTestHelper uses the wrong constructor to create the checker instance, making it impossible to pass flags in a test.

Could you provide a minimal example which is a problem?

**Comment by Stephan202** [comment]
> Could you provide a minimal example which is a problem?

IIRC ~any test that provides a non-vacuous set of flags using `setArgs` should trigger the issue.

On example:
```sh
git clone git@github.com:PicnicSupermarket/error-prone-support.git
cd error-prone-support
sed -i '/Inject/d' error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/RedundantStringConversion.java
mvn clean install -Dverification.skip -DskipTests
mvn test -Dtest='RedundantStringConversionTest#identificationOfCustomConversionMethod' -pl error-prone-contrib
```

**Comment by graememorgan** [comment]
> sed -i '/Inject/d' error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/RedundantStringConversion.java

But can it be fixed by just including the `@Inject`? Ideally we'd like `@Inject` to be used for the injectable constructor.

(I'm not suggesting not fixing this to match the original behaviour, but chucking some `@Inject`s around would be good anyway.)

**Comment by ksobolew** [comment]
Yes, it can be fixed this way. I originally didn't realize that, so the severity is much lower than I thought. Now it's just mostly a source of confusion :)

**Comment by Stephan202** [comment]
The warning-level [`InjectOnBugCheckers`](https://errorprone.info/bugpattern/InjectOnBugCheckers`) check should help, though many may have disabled it due to the false-positives it reported in an earlier version. Perhaps adding a hint to the [release notes](https://github.com/google/error-prone/releases/tag/v2.19.0) could also help. :)

**Comment by graememorgan** [comment]
Should be fixed now; give me a shout about any further problems.
